## Iris flower detection model

#### Importing necessary libraries

In [55]:
# Importing Pandas Library as 'pd'
import pandas as pd
# Importing NumPy Library as 'np'
import numpy as np

#### Load the dataset to python environment

In [56]:
# Getting data from Local Excel File and copying into iris_data Datafram
iris_data=pd.read_excel('iris (2).xls')

In [57]:
# Getting data of dataframe iris_data
iris_data.head()

SL   SW   PL   PW Classification
0  5.1  3.5  1.4  0.2    Iris-setosa
1  4.9  3.0  1.4  0.2    Iris-setosa
2  4.7  3.2  1.3  0.2    Iris-setosa
3  4.6  3.1  1.5  0.2    Iris-setosa
4  5.0  3.6  1.4  0.2    Iris-setosa

In [58]:
iris_data.shape

(150, 5)

In [59]:
iris_data.describe()

SL          SW          PL          PW
count  150.000000  150.000000  150.000000  150.000000
mean     5.843333    3.054000    3.758667    1.198667
std      0.828066    0.433594    1.764420    0.763161
min      4.300000    2.000000    1.000000    0.100000
25%      5.100000    2.800000    1.600000    0.300000
50%      5.800000    3.000000    4.350000    1.300000
75%      6.400000    3.300000    5.100000    1.800000
max      7.900000    4.400000    6.900000    2.500000

In [60]:
iris_data.describe(include=['object'])

Classification
count              150
unique               3
top     Iris-virginica
freq                50

In [61]:
iris_data.isna().sum().to_frame('Null value count')

Null value count
SL                             0
SW                             0
PL                             0
PW                             0
Classification                 0

##**Insights:**


*   The dataset contains the details of iris flower, like -          
      *   **SL** - The sepal lenght on centimeters_
      *   **SW** - The sepal width on centimeters _
      *   **PL** - The Petal lenght on centimeters _
      *   **PW** - The Petal Width on centimeters_
      *   **Classification** -The dataset have three types of iris:[ Iris-setosa,Iris-versicolour,Iris-virginica ]_
     
*   **150 Observations** and **5 Features** are found.
*   The **Null values** are recorded for:
      *   No null values found
*  The **datatype** for the features - 
      *   _**Classification** - is given as Objects_
      *   _**SL,SW,PL,PW** - are given as Float with 64-bits_
        

In [62]:
# Getting a list of column names from feature dataframe which are suitable for label encoding
Iris_data_Cols = []
for i in list(iris_data.columns):
    if iris_data[i].dtype == 'O' and len(iris_data[i].unique()) > 2:
        Iris_data_Cols.append(i)
        
print('The below given list of columns needs a Label Encoding:\n',Iris_data_Cols)

The below given list of columns needs a Label Encoding:
 ['Classification']


In [63]:
# Creating a label encoded dataframe of the targeted columns with the help of LabelEncoder from sklearn.preprocessing
from sklearn.preprocessing import LabelEncoder
Label_Enc= LabelEncoder()
temp_df=pd.DataFrame()
temp_df2=pd.DataFrame()
for k in Iris_data_Cols:
    l=iris_data[k].values
    l=Label_Enc.fit_transform(l)
    temp_df2=pd.DataFrame(l, columns=[str(k)])
    temp_df = pd.concat([temp_df, temp_df2], axis=1)

In [64]:
# Replacing the original columns in the feature dataframe with encoded columns
import warnings
warnings.filterwarnings("ignore")
iris_data[Iris_data_Cols]=temp_df[Iris_data_Cols]
# Sample preview of the feature dataframe
iris_data.sample(10)

SL   SW   PL   PW  Classification
90   5.5  2.6  4.4  1.2               1
47   4.6  3.2  1.4  0.2               0
142  5.8  2.7  5.1  1.9               2
14   5.8  4.0  1.2  0.2               0
76   6.8  2.8  4.8  1.4               1
115  6.4  3.2  5.3  2.3               2
5    5.4  3.9  1.7  0.4               0
123  6.3  2.7  4.9  1.8               2
29   4.7  3.2  1.6  0.2               0
48   5.3  3.7  1.5  0.2               0

In [65]:
X = iris_data.drop(['Classification'], axis=1)
y = iris_data.Classification

In [66]:
# splitting the data into train and test
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.3, random_state=42, shuffle=True)

In [67]:
from sklearn import preprocessing
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.model_selection import cross_val_score
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.neighbors import KNeighborsRegressor
from sklearn.ensemble import GradientBoostingRegressor

In [68]:
standardisation = preprocessing.StandardScaler()
x= standardisation.fit_transform(x)
x = pd.DataFrame(x)
x.describe()

0             1             2             3
count  1.500000e+02  1.500000e+02  1.500000e+02  1.500000e+02
mean   1.049161e-16 -9.150088e-17 -2.649732e-16  1.609823e-15
std    1.003350e+00  1.003350e+00  1.003350e+00  1.003350e+00
min   -1.870024e+00 -2.438987e+00 -1.568735e+00 -1.444450e+00
25%   -9.006812e-01 -5.877635e-01 -1.227541e+00 -1.181504e+00
50%   -5.250608e-02 -1.249576e-01  3.362659e-01  1.332259e-01
75%    6.745011e-01  5.692513e-01  7.627586e-01  7.905908e-01
max    2.492019e+00  3.114684e+00  1.786341e+00  1.710902e+00

#### 1.Logistic regression

In [76]:
# LogisticRegression
classifier = LogisticRegression()
classifier.fit(x_train,y_train)
y_pred = classifier.predict(x_test)

# Summary of the predictions made by the classifier
print('MES is : ',mean_squared_error(y_test, y_pred))
print('R squared value is:', r2_score(y_test, y_pred))
# Accuracy score
from sklearn.metrics import accuracy_score
print('accuracy is',accuracy_score(y_pred,y_test))
# f1 score
from sklearn.metrics import f1_score
print('f1 score is',f1_score(y_test, y_pred, average='macro'))

MES is :  0.0
R squared value is: 1.0
accuracy is 1.0
f1 score is 1.0


In [78]:
p=classifier.predict([[5.1,3.5,1.4,0.2]])
print(p[0])

2


#### 2. K-Nearest Neighbour

In [75]:
# K-Nearest Neighbours
#from sklearn.neighbors import KNeighborsClassifier
classifier = KNeighborsClassifier(n_neighbors=8)
classifier.fit(x_train, y_train)

y_pred = classifier.predict(x_test)

# Summary of the predictions made by the classifier
print('MES is : ',mean_squared_error(y_test, y_pred))
print('R squared value is:', r2_score(y_test, y_pred))
# Accuracy score
from sklearn.metrics import accuracy_score
print('accuracy is',accuracy_score(y_pred,y_test))
# f1 score
from sklearn.metrics import f1_score
print('f1 score is',f1_score(y_test, y_pred, average='macro'))

MES is :  0.0
R squared value is: 1.0
accuracy is 1.0
f1 score is 1.0


#### 3.Support vector machine's

In [74]:
# Support Vector Machine's 
from sklearn.svm import SVC

classifier = SVC()
classifier.fit(x_train, y_train)

y_pred = classifier.predict(x_test)

# Summary of the predictions made by the classifier
print('MES is : ',mean_squared_error(y_test, y_pred))
print('R squared value is:', r2_score(y_test, y_pred))
# Accuracy score
from sklearn.metrics import accuracy_score
print('accuracy is',accuracy_score(y_pred,y_test))
# f1 score
from sklearn.metrics import f1_score
print('f1 score is',f1_score(y_test, y_pred, average='micro'))

MES is :  0.0
R squared value is: 1.0
accuracy is 1.0
f1 score is 1.0


#### 4. Decision Tree's

In [72]:
# Decision Tree's
from sklearn.tree import DecisionTreeClassifier

classifier = DecisionTreeClassifier()

classifier.fit(x_train, y_train)

y_pred = classifier.predict(x_test)

# Summary of the predictions made by the classifier
print('MES is : ',mean_squared_error(y_test, y_pred))
print('R squared value is:', r2_score(y_test, y_pred))
# Accuracy score
from sklearn.metrics import accuracy_score
print('accuracy is',accuracy_score(y_pred,y_test))
# f1 score
from sklearn.metrics import f1_score
print('f1 score is',f1_score(y_test, y_pred, average='micro'))

MES is :  0.044444444444444446
R squared value is: 0.9358974358974359
accuracy is 0.9555555555555556
f1 score is 0.9555555555555556


#### 5.Random Forest Model

In [77]:
# Random Forest Model
#from sklearn.ensemble import RandomForestClassifier
classifier = RandomForestClassifier()
classifier.fit(x_train, y_train)

y_pred = classifier.predict(x_test)

# Summary of the predictions made by the classifier
print('MES is : ',mean_squared_error(y_test, y_pred))
print('R squared value is:', r2_score(y_test, y_pred))
# Accuracy score
from sklearn.metrics import accuracy_score
print('accuracy is',accuracy_score(y_pred,y_test))
# f1 score
from sklearn.metrics import f1_score
print('f1 score is',f1_score(y_test, y_pred, average='micro'))

MES is :  0.0
R squared value is: 1.0
accuracy is 1.0
f1 score is 1.0
